internal/venice-common/src/main/java/com/linkedin/venice/memory/HeapSizeEstimator.java

```diff
@@ -0,0 +1,188 @@
+package com.linkedin.venice.memory;
+
+import com.linkedin.venice.common.Measurable;
+import com.sun.management.HotSpotDiagnosticMXBean;
+import java.lang.management.ManagementFactory;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Utility class to help in implementing {@link Measurable#getSize()}. A couple of important points:
+ *
+ * 1. This utility class does not "measure" the heap size, but rather attempts to "predict" it, based on knowledge of
+ *    the internals of the JVM. If any of the assumptions are wrong, then of course the results will be inaccurate.
+ * 2. This utility class assumes we are using the HotSpot JVM.
+ */
+public class HeapSizeEstimator {
```

Comment:
Starting a thread to respond to @gaojieliu's general comment:

>Have you got a chance to take a look at openjdk/jol implementation?
>https://github.com/openjdk/jol
>It might carry some issues, but I guess we can analyze the strategy this lib is using to see whether we can borrow some or not.

I posted this work on social media and received a lot of interesting examples from folks about other open source projects with the same or similar purpose. I'll maintain a list of these here:

1. https://github.com/openjdk/jol
2. https://github.com/ehcache/sizeof
3. https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/RamUsageEstimator.java

For now, this work looks pretty similar to (at least some parts of) the above, but I think there is still potential to have a good implementation as part of Venice. In particular, I want a solution in which there is no hot path reflection. For now, this tradeoff is not apparent because I have not yet included examples of how we can leverage this utility in the rest of our code, but I will add it soon when the basic functionality is stabilized...

Reply:
The `reflection` mentioned here is mainly about performance, right?
Have you tried these alternatives and run it with JMH?
I tried `jol` and the initialization is slow, but after that, it is fairly fast.

Reply:
I did a small test against `jol` and it seems it can work properly with Oracle OpenJDK8 release, but not MSFT OpenJDK11 release.

Reply:
@alex-dubrouski 
Do you know why MSFT-jdk11 won't work well with attached agent to measure the object size?
I tried to add this jvm arg, but it still doesn't work:
```
-Djdk.attach.allowAttachSelf=true
```

And both `jol` and `sizeof` are suffering from the same issue with JDK11 as they are using the similar technologies..

Reply:
It works, just shows warning that dynamic attach failed:
```
 [ ~/tools/jol ]$ . ~/.j17
 [ ~/tools/jol ]$ java -cp .:jol-cli.jar org.openjdk.jol.Main internals java.lang.String
# WARNING: Unable to get Instrumentation. Dynamic Attach failed. You may add this JAR as -javaagent manually, or supply -Djdk.attach.allowAttachSelf
# WARNING: Unable to attach Serviceability Agent. sun.jvm.hotspot.memory.Universe.getNarrowOopBase()
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 3-bit shift.
# WARNING | Compressed references base/shifts are guessed by the experiment!
# WARNING | Therefore, computed addresses are just guesses, and ARE NOT RELIABLE.
# WARNING | Make sure to attach Serviceability Agent to get the reliable addresses.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

java.lang.String object internals:
OFF  SZ      TYPE DESCRIPTION               VALUE
  0   8           (object header: mark)     0x0000000000000001 (non-biasable; age: 0)
  8   4           (object header: class)    0x0000eba0
 12   4       int String.hash               0
 16   1      byte String.coder              0
 17   1   boolean String.hashIsZero         false
 18   2           (alignment/padding gap)   
 20   4    byte[] String.value              []
Instance size: 24 bytes
Space losses: 2 bytes internal + 0 bytes external = 2 bytes total

```

Reply:
What is this `String.hashIsZero` field? I don't see that when looking at the source code 🤔

Reply:
It turned out that msft-jdk11.0.8/11.0.13 has a bug, which triggered the crash while using 'jol' or 'sizeof' lib and JDK17 apparently fixed this issue.

Reply:
Well, Felix, that's why we should probably use JOL :D ;)

Reply:
But where is the learning value in that, @alex-dubrouski :D ?

Besides, this memory estimation code needs to run in DVC, which is one of our client libs, meaning we cannot tightly control which JVM it's going to run in. Ideally, it should not depend on agents or finicky `Unsafe` code...

That being said, using JOL in unit test to validate the accuracy of this utility might be a good idea 🤔

Reply:
BTW... you did not answer my question @alex-dubrouski 😅 but I found out that the `hashIsZero` was added 5 years ago: https://github.com/openjdk/jdk/commit/89a267ca46d6252e745fe8f2f1a9cb618af5f83f

According to [JDK-8221836](https://bugs.openjdk.org/browse/JDK-8221836), this change got released in Java 13, which is weird because I thought I was looking at the Java 17 source code in my IDE, but apparently not; it was probably hanging on to the source of an older Java version.

Reply:
And indeed, I can get it to show me the proper source and see that it's there now.

Reply:
I tested on JDK17 so yeah, it is there. Plus it is a field not method.